File: iosApp/iosApp/kotlin/local/AttributeSubscriber.swift
//
//  AttributeSubscriber.swift
//  iosApp
//
//  Created by Sylwester Zielinski on 23/04/2026.
//

import Matter
import SharedCode
import OSLog

struct AttributeUpdate<T : Sendable> : Sendable {
    let value: T?
    let error: Error?
}

class AttributeSubscriber {
    
    private let logger = Logger(subsystem: "nrf.matter", category: "AttributeSubscriber")
    private let baseDevice: MTRBaseDevice
    
    init(deviceId: NSNumber) throws {
        let controller = try LocalControllerProvider(logTag: "AttributeSubscriber").getController()
        baseDevice = MTRBaseDevice(nodeID: deviceId, controller: controller)
    }
    
    func subscribe<T: AttributeParser>(endpoint: NSNumber, cluster: NSNumber, attribute: NSNumber, onUpdate: @escaping (T) -> Void) {
        baseDevice.subscribeToAttributes(
            withEndpointID: endpoint,
            clusterID: cluster,
            attributeID: attribute,
            params: nil,
            queue: DispatchQueue.global(),
            reportHandler: { result, error in
                
                if let error = error {
                    // todo
                    return
                }
                
                if let result = result, let value = try? T.parse(value: result[0].readAny()) {
                    onUpdate(value)
                }
            }
        )
    }
}

File: composeApp/src/iosMain/kotlin/no/nordicsemi/nrf/matter/matter/AttributeSubscriber.kt
package no.nordicsemi.nrf.matter.matter

import kotlinx.cinterop.ExperimentalForeignApi
import platform.Foundation.NSNumber
import platform.Matter.MTRBaseDevice
import platform.Matter.MTRSubscribeParams

/**
 * Mirrors `iosApp/iosApp/kotlin/local/AttributeSubscriber.swift`.
 *
 * Unlike the Swift version, no boxing wrapper (`IosFlowWrapper`/`KotlinBoolean`) is needed:
 * that machinery existed only to cross the Swift-to-Kotlin ObjC-generics boundary. Pure
 * Kotlin-to-Kotlin, calling into a plain `MutableSharedFlow`, has no such boundary.
 */
@OptIn(ExperimentalForeignApi::class)
internal class AttributeSubscriber(deviceId: NSNumber) {

    private val baseDevice: MTRBaseDevice = MTRBaseDevice.deviceWithNodeID(
        nodeID = deviceId,
        controller = LocalControllerProviderImpl.getController("AttributeSubscriber"),
    )

    fun <T> subscribe(
        endpoint: NSNumber,
        cluster: NSNumber,
        attribute: NSNumber,
        parser: AttributeParser<T>,
        onUpdate: (T) -> Unit,
    ) {
        baseDevice.subscribeToAttributesWithEndpointID(
            endpointID = endpoint,
            clusterID = cluster,
            attributeID = attribute,
            params = defaultSubscribeParams(),
            queue = defaultMatterQueue(),
            reportHandler = { values, error ->
                if (error == null) {
                    val report = values?.firstOrNull() as? Map<*, *>
                    val data = report?.get("data") as? Map<*, *>
                    if (data != null && data.containsKey("value")) {
                        val parsed = runCatching { parser.parse(data["value"]) }
                        parsed.getOrNull()?.let(onUpdate)
                    }
                }
            },
            subscriptionEstablished = {},
        )
    }

    private fun defaultSubscribeParams(): MTRSubscribeParams =
        MTRSubscribeParams(minInterval = NSNumber(int = 0), maxInterval = NSNumber(int = 0))
}
